Encoding a struct with a sub-record field default value failed

My understanding of Avro is that fields of type ```Record``` support default values only if the sub-record is defined separately (not as a nested record).

I tried the following scenario:

```go 
func TestDebug(t *testing.T) {
	_ = avro.MustParse(`{
		"name": "service.A",
		"type": "record",
		"fields": [{
			"name": "a",
			"type": "string"
		}]
	}`)
	schema := avro.MustParse(`{
		"name": "service.B",
		"type": "record",
		"fields": [{
			"name": "a",
			"type": "service.A",
			"default": {"a": "a value"}
		},{
			"name": "b",
			"type": "string"
		}]
	}`)

	type A struct {
		A string `avro:"a"`
	}

	type B struct {
		B string `avro:"b"`
	}

	b, _ := avro.Marshal(schema, B{B: "b value"})

	type B1 struct {
		B string `avro:"b"`
		A A      `avro:"a"`
	}

	b1 := B1{}
	_ = avro.Unmarshal(schema, b, &b1)

	if want, got := (A{A: "a value"}), b1.A; want != got {
		t.Errorf("%v,%v must be equals", want, got)
	}
}

```


<details>
<summary>It panics with the STACK TRACE (unfold)</summary>

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x100ee7b]

goroutine 6 [running]:
testing.tRunner.func1.2({0x11c9020, 0x137f0e0})
        /usr/local/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1548 +0x397
panic({0x11c9020?, 0x137f0e0?})
        /usr/local/go/src/runtime/panic.go:914 +0x21f
reflect.mapaccess(0xc00004e460?, 0xc00010d930?, 0xc00010d9b8?)
        /usr/local/go/src/runtime/map.go:1350 +0x13
github.com/modern-go/reflect2.(*UnsafeMapType).UnsafeGetIndex(...)
        /Users/mac/go/pkg/mod/github.com/modern-go/reflect2@v1.0.2/unsafe_map.go:101
github.com/hamba/avro/v2.(*recordMapEncoder).Encode(0xc00006a880, 0xc00010be30, 0xc000029040)
        /Users/mac/dev/hamba_avro/codec_record.go:337 +0x107
github.com/hamba/avro/v2.(*structEncoder).Encode(0x13882b0?, 0xc000027dc0, 0xc000029040)
        /Users/mac/dev/hamba_avro/codec_record.go:204 +0xa7
github.com/hamba/avro/v2.(*Writer).WriteVal(0xc000029040, {0x124ce48, 0xc00011c370}, {0x11cff60, 0xc000027dc0})
        /Users/mac/dev/hamba_avro/codec.go:63 +0xed
github.com/hamba/avro/v2.(*frozenConfig).Marshal(0xc00012c080, {0x124ce48, 0xc00011c370}, {0x11cff60, 0xc000027dc0})
        /Users/mac/dev/hamba_avro/config.go:126 +0x5a
github.com/hamba/avro/v2.Marshal(...)
        /Users/mac/dev/hamba_avro/encoder.go:36
github.com/ln80/event-store/avro/v2.TestDebug(0xc0001269c0)
        /Users/mac/dev/larcade/event-store/avro/v2/draft_test.go:355 +0xd3
testing.tRunner(0xc0001269c0, 0x12146d0)
        /usr/local/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
        /usr/local/go/src/testing/testing.go:1648 +0x3ad
```
</details>


I dug deeper into the code, and it seems to me that ``` *reflect2.UnsafeMapType.UnsafeGetIndex ``` receives the wrong argument:

https://github.com/hamba/avro/blob/58cadd08715cb07eb2fa3e19b1a31bb0c0fbd96d/codec_record.go#L328-L332

``ptr`` should be an unsafe pointer of a map’s pointer: ```unsafe.Pointer(&map[string]any)```

--- a/codec_record.go
+++ b/codec_record.go
@@ -172,9 +172,14 @@ func encoderOfStruct(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValEnc
 		}
 
 		defaultType := reflect2.TypeOf(def)
+		defaultEncoder := encoderOfType(cfg, field.Type(), defaultType)
+		if defaultType.LikePtr() {
+			defaultEncoder = &onePtrEncoder{defaultEncoder}
+		}
+
 		fields = append(fields, &structFieldEncoder{
 			defaultPtr: reflect2.PtrOf(def),
-			encoder:    encoderOfType(cfg, field.Type(), defaultType),
+			encoder:    defaultEncoder,
 		})
 	}
 	return &structEncoder{typ: typ, fields: fields}

--- a/encoder_record_test.go
+++ b/encoder_record_test.go
@@ -139,6 +139,42 @@ func TestEncoder_RecordStructPartialWithNullDefault(t *testing.T) {
 	assert.Equal(t, []byte{0x00, 0x06, 0x66, 0x6f, 0x6f}, buf.Bytes())
 }
 
+func TestEncoder_RecordStructPartialWithSubRecordDefault(t *testing.T) {
+	defer ConfigTeardown()
+
+	_, err := avro.Parse(`{
+		"type": "record",
+		"name": "test",
+		"fields" : [
+			{"name": "a", "type": "long"},
+			{"name": "b", "type": "string"}
+		]
+	}`)
+	require.NoError(t, err)
+
+	schema := `{
+		"type": "record",
+		"name": "parent",
+		"fields" : [
+			{
+				"name": "a",
+				"type": "test",
+				"default": {"a": 1000, "b": "def b"}
+			},
+			{"name": "b", "type": "string"}
+		]
+	}`
+	obj := TestPartialRecord{B: "foo"}
+	buf := &bytes.Buffer{}
+	enc, err := avro.NewEncoder(schema, buf)
+	require.NoError(t, err)
+
+	err = enc.Encode(obj)
+	require.NoError(t, err)
+
+	assert.Equal(t, []byte{0xd0, 0xf, 0xa, 0x64, 0x65, 0x66, 0x20, 0x62, 0x6, 0x66, 0x6f, 0x6f}, buf.Bytes())
+}
+
 func TestEncoder_RecordStructWithNullDefault(t *testing.T) {
 	defer ConfigTeardown()
 
@@ -408,6 +444,43 @@ func TestEncoder_RecordMapWithDefault(t *testing.T) {
 
 	require.NoError(t, err)
 	assert.Equal(t, []byte{0x36, 0x06, 0x66, 0x6f, 0x6f}, buf.Bytes())
+}
+
+func TestEncoder_RecordMapWithSubRecordDefault(t *testing.T) {
+	defer ConfigTeardown()
+
+	_, err := avro.Parse(`{
+		"type": "record",
+		"name": "test",
+		"fields" : [
+			{"name": "a", "type": "long"},
+			{"name": "b", "type": "string"}
+		]
+	}`)
+	require.NoError(t, err)
+
+	schema := `{
+		"type": "record",
+		"name": "parent",
+		"fields" : [
+			{
+				"name": "a",
+				"type": "test",
+				"default": {"a": 1000, "b": "def b"}
+			},
+			{"name": "b", "type": "string"}
+		]
+	}`
+
+	obj := map[string]any{"b": "foo"}
+	buf := &bytes.Buffer{}
+	enc, err := avro.NewEncoder(schema, buf)
+	require.NoError(t, err)
+
+	err = enc.Encode(obj)
+	require.NoError(t, err)
+
+	assert.Equal(t, []byte{0xd0, 0xf, 0xa, 0x64, 0x65, 0x66, 0x20, 0x62, 0x6, 0x66, 0x6f, 0x6f}, buf.Bytes())
 }
 
 func TestEncoder_RecordMapWithNullDefault(t *testing.T) {

